PEP 747: TypeExpr: Type Hint for a Type Expression

davidfstr opened the pull request:
Also:
* ~~PEP 12: Explain how to use Intersphinx references. Mention that Typing docs can be referenced using them.~~
* PEP 655: Remove gremlin character

CC sponsor @JelleZijlstra 

<!--
You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.

If your PEP is not Standards Track, remove the corresponding section.
-->

## Basic requirements (all PEP Types)

* [x] Read and followed [PEP 1](https://peps.python.org/1) & [PEP 12](https://peps.python.org/12)
* [x] File created from the [latest PEP template](https://github.com/python/peps/blob/main/peps/pep-0012/pep-NNNN.rst?plain=1)
* [x] PEP has next available number, & set in filename (``pep-NNNN.rst``), PR title (``PEP 123: <Title of PEP>``) and ``PEP`` header
* [x] Title clearly, accurately and concisely describes the content in 79 characters or less
* [ ] Core dev/PEP editor listed as ``Author`` or ``Sponsor``, and formally confirmed their approval
    * DF: @JelleZijlstra has agreed to sponsor. Is that the approval required?
* [x] ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
* [x] ``PEP-Delegate``, ``Topic``, ``Requires`` and ``Replaces`` headers completed if appropriate
* [x] Required sections included
    * [x] Abstract (first section)
    * [x] Copyright (last section; exact wording from template required)
* [x] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer names](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
    * DF: All code blocks are Python. Hopefully they are all PEP 8 compliant. I don't know how to check using a tool.
* [x] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
    * ~~DF: There is a lint error on the `Post-History` line which I don't know how to fix~~
* [x] Authors/sponsor added to ``.github/CODEOWNERS`` for the PEP


## Standards Track requirements

* [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate
* [x] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [x] Motivation
    * [x] Rationale
    * [x] Specification
    * [x] Backwards Compatibility
    * [#] Security Implications
    * [x] How to Teach This
    * [x] Reference Implementation
    * [x] Rejected Ideas
    * [#] Open Issues
* [x] ``Python-Version`` set to valid (pre-beta) future Python version, if relevant
* [x] Any project stated in the PEP as supporting/endorsing/benefiting from the PEP formally confirmed such
    * DF: No endorsements included.
    * DF: However section §"Common kinds of functions that would benefit from TypeExpr" mentions several libraries that are expected to benefit from this PEP. Maintainers for all those libraries were [tagged on a GitHub discussion thread](https://github.com/python/mypy/issues/9773#issuecomment-2094360216). Only maintainer Hynek (of the attrs and svcs libraries) said there would be no benefit, and those libraries are no longer mentioned in this latest PEP draft.
* [ ] Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in ``Discussions-To`` and ``Post-History``
    * ~~DF: After merging I plan to post to [this existing thread], then make a small PR to link to the exact post.~~
    * DF: After merging I plan to post to a new thread, then make a small PR linking to the thread in both Discussions-To and Post-History.

[this existing thread]: https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435

<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--3798.org.readthedocs.build/

<!-- readthedocs-preview pep-previews end -->

davidfstr committed:
PEP 755: TypeExpr: Type Hint for a Type Expression

Also:
* PEP 655: Remove gremlin character

hugovk commented on .github/CODEOWNERS:
```diff
@@ -626,6 +626,7 @@ peps/pep-0744.rst  @brandtbucher
 peps/pep-0745.rst  @hugovk
 # ...
 # peps/pep-0754.rst
+peps/pep-0755.rst  @davidfstr @JelleZijlstra
```
Only those with write access can be in the CODEOWNERS:
```suggestion
peps/pep-0755.rst  @JelleZijlstra
```

hugovk commented on peps/pep-0755.rst:
```diff
@@ -0,0 +1,907 @@
+PEP: 755
```
PEP 747 is the next available, please can you use that?

davidfstr replied:
Will adjust.

hugovk commented on peps/pep-0755.rst:
```diff
@@ -0,0 +1,907 @@
+Title: TypeExpr: Type Hint for a Type Expression
+Author: David Foster <david at dafoster.net>
+Sponsor: Jelle Zijlstra <jelle.zijlstra at gmail.com>
+Discussions-To: https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435
```
_After_ the PEP has been merged into the repo, this will be the new discussion link.

davidfstr replied:
Should I remove the Discussions-To header until the initial merge is complete? Or is there something else I should do with this header?

davidfstr replied:
Rereading PEP 1 carefully, I see:

> As soon as a PEP number has been assigned and the draft PEP is committed to the [PEP repository](https://github.com/python/peps), a discussion thread for the PEP should be created to provide a central place to discuss and review its contents, and the PEP should be updated so that the Discussions-To header links to it.

This seems to imply that the original draft PEP must not include a Discussions-To header and it must not include a Post-History. Both of those headers should be created immediately after a _new_ thread for discussing the PEP is created.

So it sounds like I should:
* Delete the Discussions-To and Post-History headers for now, since there's no thread to link to yet.
* Shepherd this PR to merge.
* Create a new discussion thread. Link the old thread to the new thread. Link the PEP to the new thread via a small change adding a Discussions-To and Post-History header.

**Does that sound right** @hugovk / @JelleZijlstra ?

There's already a pretty substantial [discussion thread for drafts 1 and 2](https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435) of this PEP (before the current draft 3 in this PR). **Should a link to this old thread be (eventually) included in Discussions-To as well?**

JelleZijlstra replied:
Yes, that's right. The existing thread should be listed in Post-History.

hugovk commented on peps/pep-0755.rst:
```diff
@@ -0,0 +1,907 @@
+Topic: Typing
+Created: 21-Dec-2020
+Python-Version: 3.14
+Post-History: 24-Jan-2021, 28-Jan-2021, 04-Feb-2021, 14-Feb-2021, `19-Apr-2024 <https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435>`__, `04-May-2024 <https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/7>`__
```
And this will also be the Discussions-To link, and any new ones opened to discuss the PEP later.

hugovk commented on peps/pep-0755.rst:
```diff
@@ -0,0 +1,907 @@
+Abstract
+========
+
+PEP 484 defines the notation ``type[C]`` where ``C`` is a class, to
```
Use Sphinx role here, and similar ones below:
```suggestion
:pep`484` defines the notation ``type[C]`` where ``C`` is a class, to
```

hugovk commented on peps/pep-0755.rst:
```diff
@@ -0,0 +1,907 @@
+Subtyping
+---------
+
+Whether a TypeExpr value can be assigned from one variable to another is
```
Consistently apply code formatting for things like this throughout the document:
```suggestion
Whether a ``TypeExpr`` value can be assigned from one variable to another is
```

hugovk commented:
> * [ ] Core dev/PEP editor listed as Author or Sponsor, and formally confirmed their approval
>   * DF: @JelleZijlstra has agreed to sponsor. Is that the approval required?

Yes, Jelle will confirm in this PR.

> * [ ] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
>   * DF: There is a lint error on the `Post-History` line which I don't know how to fix

`Discussions-To` should be the _new thread_ opened for the PEP draft. It also goes in `Post-History`. If you have further updates later which warrant a new thread, this is put in `Discussions-To`, and both threads are listed in `Post-History`. We don't need to list all the pre-PEP discussions.


> * [ ]  Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in `Discussions-To` and `Post-History`
>   
>   * DF: After merging I plan to post to [this existing thread](https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435), then make a small PR to link to the exact post.

Please open a new thread after the PEP is merged to the repo. That gives notice that there's a new PEP draft to be reviewed, and will make it more widely known: there are people not following that existing thread who will pay more attention to a new thread announcing a new PEP draft.

hugovk commented on peps/pep-0755.rst:
```diff
@@ -0,0 +1,907 @@
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Created: 21-Dec-2020
```
[PEP 1](https://peps.python.org/pep-0001/#pep-header-preamble) says:

> The Created header records the date that the PEP was assigned a number [...]

davidfstr committed:
Update .github/CODEOWNERS

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

davidfstr committed:
Update Created header

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

davidfstr committed:
Post-History: Recognize discuss.python.org post links without trailing slash

Previously a URL like the following was recognized:
    https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/7/
But the canonical URL actually has no trailing slash:
    https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/7
That canonical URL is now recognized.

davidfstr committed:
Include only draft 1 and 2 links in Post-History. Remove Discussions-To until draft 3 link available.

davidfstr committed:
Change PEP number: 755 -> 747

davidfstr committed:
Rewrite some PEP links using :pep:`...`

davidfstr committed:
Add code formatting in many places

davidfstr commented:
Feedback applied:
* PEP number changed.
* Post-History trimmed to point to draft 1 and draft 2 discussions only, and not any prior discussions.
* Discussions-To removed until this draft 3 is shepherded to merge and a new discussion link for draft 3 is created.
* Code formatting applied in many places.

One interesting thing I had to do was [extend the linting infrastructure slightly](https://github.com/python/peps/pull/3798/commits/0bbbd9c58d0dbfb125434edd870d6cdf88fbaa96) to recognize discuss.python.org post links without a trailing slash (which is the format the web UI generates the links in by default).

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+
+:pep:`484` defines the notation ``type[C]`` where ``C`` is a class, to
+refer to a class object that is a subtype of ``C``. It explicitly does
+not allow ``type[C]`` to refer to arbitrary `type expression`_ objects such
```
```suggestion
not allow ``type[C]`` to refer to arbitrary :ref:typing:`type expression <type-expression>` objects such
```
This allows to remove the footer links below.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+   -  Examples: attrs.\ `make_class`_, dataclasses.\ `make_dataclass`_, `openapify`_
+
+.. _make_class: https://www.attrs.org/en/stable/api.html#attrs.make_class
+.. _make_dataclass: https://github.com/python/typeshed/issues/11653
```
Wouldn't this need an annotation expression instead of a type expression?

davidfstr replied:
I don't think so. None of the "annotation expression but not type expression" forms - with 1 exception† - make sense to use as field types: (Not)Required, ReadOnly, ClassVar, Final, InitVar† (🤔), Unpack.

† InitVar is the only problematic case.

I'll probably need to think about InitVar some more...

davidfstr replied:
I added a footnote talking about `InitVar` being a special case with `dataclasses.make_dataclass`

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+Specification
+=============
+
+A ``TypeExpr`` represents a `type expression`_ such as ``Optional[str]``,
+``Union[int, str]``, or ``MyTypedDict``. A ``TypeExpr`` type is written as
```
```suggestion
A ``TypeExpr`` represents a `type expression`_ such as ``str | None``,
``dict[int, str]``, or ``MyTypedDict``. A ``TypeExpr`` type is written as
```
Let's not lead with a deprecated object.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+checker should apply its usual type inference mechanisms to determine
+the type of its argument, possibly ``Any``.
+
+Type expressions include more values than `annotation expressions`_ but
```
For the benefit of readers of the PEP, it may be worth recapitulating what a "type expression" is. I think that may fit well in a "Rationale" section, which is about justifying why we're using this precise approach.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+Using TypeExprs
+---------------
+
+A ``TypeExpr`` type may be used in any context where a type annotation is
```
This can be written more concisely by saying that ``TypeExpr`` itself is a new kind of type expression, which implies that it is valid anywhere a type expression is expected. The term "type annotation" in this line should at least be rewritten to "type expression". `TypeExpr` should also be valid in other type expression contexts that are not annotations, such as in `cast()`.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+Stringified TypeExprs
+'''''''''''''''''''''
+
+A ``TypeExpr`` value may be a string literal that spells a forward
```
It should be made clear that this section does not spell out a special case; it's a natural consequence of the definitions of "type expression" and TypeExpr.

davidfstr replied:
I rephrased as:

> A ``TypeExpr``, **as a consequence of being type expression**, may be a string literal that ...

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+::
+
+   typx: TypeExpr = TypeIs[str]
+   assert_type(typx, TypeExpr[Any])  # acceptable
```
This seems like an odd special case. Why is it necessary?

davidfstr replied:
The whole concept of allowing non-universal type expressions to parameterize/constrain a type variable doesn't make a lot of sense to me, so I'm attempting to give type checkers leeway to immediately convert non-universal type expressions to `Any`, both immediately (in this section) and when in combination with type variables (in §"Interactions with type variables").

Do you think it would be better if I simply don't mention non-universal type expressions at all and leave their behavior unspecified, by omission?

JelleZijlstra replied:
I think it's fine to leave this undefined.

davidfstr replied:
I've now edited out any specification of _behavior_ of non-universal type expressions, but still left in the _definition_ of a  "non-universal type expression" since I expect it may be a useful concept to be able to name.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+::
+
+   value = as_instance(TypeIs[str])
+   assert_type(value, Any)  # acceptable
```
Type checkers should at least infer `Any | None` (which is not equivalent to Any).

davidfstr replied:
Ah. I assumed `Any | None` *was* equivalent to `Any`. Will adjust.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+Interactions with type variables
+--------------------------------
+
+``TypeExpr[]`` can constrain a type variable that is used elsewhere within
```
I'm not sure constraining is the right word here; "constraints" mean something else in the context of type variables.

AlexWaygood replied:
"can be parameterised by" might be better than "can constrain"

davidfstr replied:
If possible I'm trying to use a word that preserves the intuition that type variables used in a signature are effectively related together like an equation, and are probably "solved" using some kind of "constraint solver" (I assume).

* ~~Maybe "bind", as in "TypeExpr[] can be bound to a type variable"?~~ But then again type variables also have their own "bounds", which is something different.
* Maybe "link", as in "TypeExpr[] can be linked to a type variable"?
* Maybe "tie", as in "TypeExpr[] can be tied to a type variable"?

Of course I could fall back on the more-abstract:

* "Parameterize", as in "TypeExpr[] can be parameterized by a type variable".

Opinions on the term to use?

JelleZijlstra replied:
I think "parameterize" is actually the right term.

davidfstr replied:
I replaced "constrain" -> "parameterized by"

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+   assert_type(value, Any)  # acceptable
+
+
+Interactions with type[]
```
I'm not sure these sections need to be part of the specification. They do not appear to specify any behavior that does not follow from the rest of the specification. They might fit better in a separate section of examples, though that's not part of the standard PEP template.

davidfstr replied:
> [These sections] do not appear to specify any behavior that does not follow from the rest of the specification.

Then I'm doing a good job being consistent and not introducing surprises 😁

Still, point taken: **It may be worth moving most of the interactions that (I think) follow naturally to it's own supersection.** Some of those interactions I think are important to spell out because they show that TypeExprs can be valuable in real code. In particular I definitely want to spell out interactions with TypeIs[] and TypeGuard[] in particular.

I'll look at doing a little reorganizing.

JelleZijlstra replied:
PEPs have a dual purpose: to explain what the idea is, and to convince the reader that it's a good idea. The two naturally go together, but ideally the "Specification" section should be purely about the first part.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+   else:
+       assert_type(count, int | str)
+
+The above behavior of stripping the metadata argument to ``Annotated[]`` is
```
But type checkers such as mypy don't preserve Annotated at all, even when used directly to annotate a variable (`x: PositiveInt`). I don't see why TypeExpr is special here or why this behavior needs to be specified at all.

davidfstr replied:
In early discussions around TypeExpr (then: TypeForm) there seemed to be some concern - from the runtime type checker folks I think - that `Annotated[...]` metadata might be stripped out early due to this PEP. But that's actually the status quo anyway.

Rereading the entire §"Interactions with Annotated[]" section again, I don't think it says anything useful. Improving how `Annotated[]` interacts with the rest of the type system - and in particular preserving metadata under certain circumstances - is outside the scope of this PEP. I think I'll delete the section entirely.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+Interactions with isinstance() and issubclass()
+-----------------------------------------------
+
+The ``TypeExpr`` special form is not a ``type`` at runtime. It is
+however an ``object``:
+
+::
+
+   >>> isinstance(TypeExpr, type)
+   False
+
+   >>> isinstance(TypeExpr, object)
+   True
```
I don't think we need to specify how TypeExpr is implemented at runtime. I would remove this sentence.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+The following signatures related to class objects continue to use
+``type`` and remain unchanged:
+
+-  ``builtins.isinstance(object: object, classinfo: type | tuple[type, ...]) -> bool``
```
This is not precisely the signature of `isinstance()` (see typeshed for the full version). I'm not sure we need to spell out the signature at all; we can merely list the name of the functions.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+---------------------------------------------
+
+``type`` was `designed`_ to only be used to describe class objects. A
+class object can always be instantiated by calling it and can always be
```
This is not true; there are classes that cannot be instantiated.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+   ) -> T:
+
+However ``TypeExpr[]`` already treats :ref:`Annotated[T, metadata] as 
+equivalent to T anyway <interactions_with_annotated>`.
```
broken link

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+No backward incompatible changes are made by this PEP.
+
+
+How to Teach This
```
A question that came up at the Typing Summit that isn't explicitly addressed here: How is the *body* of a function that takes a `TypeExpr` parameter type checked? The answer is that it isn't type checked any more than if you had taken an `object` parameter, since static type checkers don't know about the runtime representation of different TypeExprs. I think it's worth spelling that out explicitly in the PEP.

JelleZijlstra reviewed (commented):
Thanks! This is a great writeup but I feel in places it goes into too much detail and specifies too many special cases.

hugovk commented:
> One interesting thing I had to do was [extend the linting infrastructure slightly](https://github.com/python/peps/pull/3798/commits/0bbbd9c58d0dbfb125434edd870d6cdf88fbaa96) to recognize discuss.python.org post links without a trailing slash (which is the format the web UI generates the links in by default).

> Previously a URL like the following was recognized:
>     `https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/7/`
> But the canonical URL actually has no trailing slash:
>     `https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/7`
> That canonical URL is now recognized.


The linter was working as expected: it's validating a `<Thread URL>`, which is meant to link to a discussion _thread_, not a _message_ that's part of a thread.

It would happily accept thread URLs such as these, with or without a trailing slash:

* `https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/`
* `https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435`

Please can you revert the linter change and link to a thread URL?

AlexWaygood commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+to a different representation at runtime:
+
+-  For example string-based forward references are normalized at runtime
+   to be ``ForwardRef`` instances: [#forward_ref_normalization]_
```
```suggestion
   to be ``ForwardRef`` instances in some contexts: [#forward_ref_normalization]_
```

AlexWaygood commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+:ref:`constraining type variables <interactions_with_type_variables>`.
+
+
+Static vs. Runtime Representations of TypeExprs
```
This section overall feels like it could be more concise. I'm not sure it's necessary to have both of your first two examples here.

AlexWaygood commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+   INT_TREE: TypeExpr = ForwardRef('IntTree')  # ERROR: Runtime-only form
+   LIST_OF_INT: TypeExpr = types.GenericAlias(list, int)  # ERROR: Runtime-only form
+
+However runtime type checkers must still be prepared to recognize and
```
"must" is quite strong language here. It almost sounds like you're mandating behaviour for other runtime type checkers. Maybe "must often"?

davidfstr replied:
Well, a runtime type checker is by definition working with the _runtime_ representations of any type expressions that it supports. This is not a new requirement on runtime type checkers; it just isn't a requirement for static type checkers.

davidfstr replied:
I think I'll replace "must" with "already must", to be clear that I'm not imposing a _new_ requirement.

AlexWaygood commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+
+``TypeExpr[]`` is covariant in its argument type, just like ``type[]``:
+
+-  ``TypeExpr[T1]`` is a subtype of ``TypeExpr[T2]`` iff ``T1`` is a
+   subtype of ``T2``.
```
Sometimes the answer to the question "Is `T1` a subtype of `T2`?" is neither "yes" or "no", but "that question doesn't make sense". (E.g., "Is `Self` a subtype of `TypeGuard[str]`?") Maybe this language might be slightly better?

```suggestion
-  ``TypeExpr[T1]`` is a subtype of ``TypeExpr[T2]`` iff ``T1`` can be
   considered a subtype of ``T2``.
```

davidfstr replied:
I don't follow: "Is X a subtype of Y?" is a statement with either a yes or no answer. If you try to assign a value of type X to a variable of type Y, either a type checker will give an error or it will not.

> Is Self a subtype of TypeGuard[str]?

The answer to this question (involving those pesky _non-universal type expressions_) I don't believe is defined by any PEP or the typing specification, so each type checker makes its own decision on the answer to the question.

I imagine the sensible answer is "no".

AlexWaygood reviewed (commented):
Thanks for writing this PEP!

davidfstr committed:
PEP 12: Extract Intersphinx section. Add example of Intersphinx reference.

davidfstr committed:
Rewrite links to Typing specification to use Intersphinx references

davidfstr committed:
Change Optional[str] example in introductions to (str | None)

davidfstr committed:
Revert "Post-History: Recognize discuss.python.org post links without trailing slash"

This reverts commit 0bbbd9c58d0dbfb125434edd870d6cdf88fbaa96.

davidfstr committed:
Alter draft 2 URL to non-canonical form to pass lint check

davidfstr commented:
> The linter was working as expected: it's validating a `<Thread URL>`, which is meant to link to a discussion _thread_, not a _message_ that's part of a thread.
> 
> It would happily accept thread URLs such as these, with or without a trailing slash:
> 
> * `https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/`
> * `https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435`
> 
> Please can you revert the linter change and link to a thread URL?

I can't change the URL of where draft 2 was introduced in the past, in the middle of an existing thread. Rounding that URL to point at the thread itself makes it more difficult to see where the draft was introduced, since the top of that thread is where draft 1 came from.

For now I've reverted the linter change and altered the draft 2 URL to point to the non-canonical URL that continues to reference exactly where draft 2 originated.

davidfstr committed:
Using TypeExprs: Clarify that TypeExpr is itself a type expression

davidfstr committed:
Remove some details about how TypeExpr is implemented at runtime

davidfstr committed:
Affected signatures in the standard library: Elide the full signature of referenced functions

davidfstr committed:
Clarify that class objects can usually, but not always, be instantiated

davidfstr committed:
Stringified TypeExprs: Clarify that behavior follows from TypeExpr being a type expression

davidfstr committed:
Clarify that (Any | None) is not the same as (Any)

davidfstr committed:
Reformat Annotated link to be all on one line

davidfstr committed:
Formatting

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

davidfstr committed:
Clarify statement about ForwardRefs

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

davidfstr committed:
Formatting

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

davidfstr committed:
Wordsmithing

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

davidfstr commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+   BadUnion2: TypeExpr = 'Union'  # ERROR: does not spell a type
+
+
+Literal[] TypeExprs
+'''''''''''''''''''
+
+To simplify static type checking, a ``Literal[...]`` value is *not*
+considered assignable to a ``TypeExpr`` variable even if all of its members
+spell valid types:
+
+::
+
+   STRS_TYPE_NAME: Literal['str', 'list[str]'] = 'str'
+   STRS_TYPE: TypeExpr = STRS_TYPE_NAME  # ERROR: Literal[] value is not a TypeExpr
+
+However ``Literal[...]`` itself is still a ``TypeExpr``:
+
+::
+
+   DIRECTION_TYPE: TypeExpr = Literal['left', 'right']  # OK
```
Eric Traut (from pyright) advocated that this particular special case be called out, so I think I'll probably still keep it.

However I can skip elaborating the "obvious" case starting with the statement:

> However ``Literal[...]`` itself is still a ``TypeExpr``:

davidfstr committed:
Literal: Remove low-value example

davidfstr committed:
Non-Universal Type Expressions: Leave behaviors undefined

davidfstr committed:
"constrain" -> "parameterized by"

davidfstr commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,903 @@
+
+::
+
+   request_json: object = ...
+   if isassignable(request_json, Shape):
+       assert_type(request_json, Shape)  # type is narrowed!
```
I'll rename `Shape` to `MyTypedDict` to hopefully be more obvious.

davidfstr committed:
Shape -> MyTypedDict

davidfstr committed:
"must" -> "already must"

hugovk commented:
> PEP 12: Explain how to use Intersphinx references. Mention that Typing docs can be referenced using them.

Please can you move the PEP 12 changes to their own PR? They'll need to be reviewed separately.

davidfstr committed:
"constrain" -> "parameterized by" (Take 2)

davidfstr committed:
Add Rationale section, explaining what a type expression is and why it's used

davidfstr committed:
Non-Universal TypeExprs: Remove explanations of where Self, TypeGuard, and TypeIs are (not) valid - which aren't helpful

davidfstr committed:
Static vs. Runtime Representations of TypeExprs: Streamline

Specifically:
* Removed 1 of 2 examples
    - The extra example wasn't necessary to illustrate the point
* Removed a paragraph which looked like an extra requirement on
  runtime checkers, which isn't the focus of this PEP

davidfstr committed:
Subtyping: Fix missing formatting for "object"

davidfstr committed:
Interactions with Annotated[]: Remove section

davidfstr committed:
How to Teach This: Rewrite section, integrating some prior 'Interactions with X' sections

davidfstr committed:
Revert "PEP 12: Extract Intersphinx section. Add example of Intersphinx reference."

This reverts commit 1c6a7af738a2588b13f7242f6595f4772faadf6a.

davidfstr committed:
Mention that dataclasses.make_dataclass() specially handles InitVar

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,856 @@
+::
+
+   count: PositiveInt = 1
+   assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
```
I don't think we should prescribe that type checkers *can't* preserve the metadata.
```suggestion
   reveal_type(count)  # either "int" or "Annotated[int, ValueRange(1, float('inf'))]"
```

davidfstr replied:
(This comment applies to the "Interactions with Annotated[]" section, which has been deleted entirely from the latest draft.)

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,856 @@
+   assert_type(count, int)  # NOT: Annotated[int, ValueRange(1, float('inf'))]
+
+In particular when an ``Annotated[]`` argument is passed to a ``TypeExpr[]``
+parameterized by a type variable that is also used by a ``TypeIs[]``
```
Hm, this actually becomes problematic with `TypeIs`, because whether `Annotated` is preserved is going to affect whether the `else` type still includes `int` in your example. I think it's OK to leave that unspecified though.

davidfstr replied:
(This comment applies to the "Interactions with Annotated[]" section, which has been deleted entirely from the latest draft.)

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,856 @@
+-  Recursive types like ``IntTree = list[typing.Union[int, 'IntTree']]``
+   are not possible to fully resolve.
+-  Supporting user-defined generic types (like Django’s
+   ``QuerySet[User]``) require user-defined functions to
```
```suggestion
   ``QuerySet[User]``) requires user-defined functions to
```
Though I'm not sure I fully understand this. Does "user-defined generic" here mean they're not using `typing.Generic`?

davidfstr replied:
"User-defined generic" here means that `typing.Generic` is being used.

However many in-the-wild uses of `typing.Generic` don't preserve enough information to check assignability to a generic type at runtime. For example, consider the following type:

```python
class Cell(Generic[T]):
    value: T
    def __init__(self, value: T) -> None:
        self.value = value
```

Given the above definition, how should a general runtime type checker answer the question `isassignable(Cell(1), Cell[int])` at runtime? The type T has been entirely erased, so it's not possible to determine for a given `Cell` instance what value for `T` was given (or inferred) statically.

So a runtime type checker must provide a user-registered hook to recognize `Cell` instances specifically and guess the cell type based on the current value of `.value` for a `Cell`.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,856 @@
+::
+
+   @overload
+   def checkcast(typx: TypeExpr[AT=Annotated[T, *Anns]], value: str) -> T: ...
```
This is a syntax error, can we just write it without the syntax error?
```suggestion
   def checkcast(typx: TypeExpr[Annotated[T, *Anns]], value: str) -> T: ...
```

davidfstr replied:
This is intended to be an example of pattern matching syntax which I'm rejecting.

I'll rephrase `For example:` as `Consider the following possible pattern matching syntax:` to make it more clear that this is not intended to be valid syntax based on what the PEP is proposing.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,856 @@
+   @overload
+   def checkcast(typx: TypeExpr[AT=Annotated[T, *Anns]], value: str) -> T: ...
+   @overload
+   def checkcast(typx: TypeExpr[UT=Union[*Ts]], value: str) -> Union[*Ts]: ...
```
```suggestion
   def checkcast(typx: TypeExpr[Union[*Ts]], value: str) -> Union[*Ts]: ...
```

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,856 @@
+::
+
+   def checkcast(
+       typx: TypeExpr[T] | TypeExpr[AT=Annotated[T, *Anns]],
```
```suggestion
       typx: TypeExpr[T] | TypeExpr[Annotated[T, *Anns]],
```

davidfstr committed:
Explicit/Implicit TypeExpr Values: Add sections

Also:
* Stringified TypeExprs: Remove now-redundant section

davidfstr committed:
Non-Universal TypeExprs: Inline section to initial 'TypeExpr Values' introduction

davidfstr committed:
Wordsmith

davidfstr commented:
(I'm still integrating feedback, including larger feedback that involves creating/reorganizing entire sections. I expect to post a new major revision to this PR in the next few days, with a comment explaining major changes made.)

davidfstr committed:
Backwards Compatibility: Mention new rules for recognizing type expressios in value expression contexts

davidfstr committed:
Motivation: Optionals -> unions

davidfstr committed:
Remove mentions to the is-consistent-with relationship, which is being redefined externally

davidfstr committed:
Support pattern matching on type expressions: Clarify examples use a proposed but rejected syntax

davidfstr committed:
Grammar

davidfstr commented:
I've pushed a new draft of this PEP, with the following major changes:

* **Rationale**: Section added. Explains what a "type expression" is, how it relates to similar concepts like "annotation expression" and "class object", and why TypeExpr was chosen to align with the definition of "type expression".
* **Specification** > **TypeExpr Values**: Added specific rules for recognizing TypeExprs in value expression contexts, based on feedback from Eric. Added the explicit `TypeExpr(...)` syntax (which parentheses).
* **Specification**: Deleted redundant subsections about **Interactions with X**, or moved/reintegrated these subsections as examples inside the **How to Teach This** supersection.
* **How to Teach This** > **Introspecting TypeExpr Values**: Added section about writing the body of a function that takes TypeExpr as input. Clarified that `TypeExpr` is treated nearly the same as `object`.

I've also reverted the PEP 12 changes in this PR, in favor of moving those changes to [a different PR](https://github.com/python/peps/pull/3827), so **brettcannon** and **warsaw** shouldn't need to stay subscribed here.

davidfstr commented:
Friendly ping to @JelleZijlstra to take a look at some of [the new major changes](https://github.com/python/peps/pull/3798#issuecomment-2156133913) in this PR for TypeExpr

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,977 @@
+
+   def trycast[T](typx: TypeExpr[T], value: object) -> T | None: ...
+
+The use of ``TypeExpr[]`` and the type variable ``T`` enable the return
```
```suggestion
The use of ``TypeExpr[]`` and the type variable ``T`` enables the return
```

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,977 @@
+That ``isassignable`` function enables a kind of enhanced ``isinstance``
+check which is useful for `checking whether a value decoded from JSON
+conforms to a particular structure`_ of nested ``TypedDict``\ s,
+``List``\ s, unions, ``Literal``\ s, and other types. This kind
```
```suggestion
lists, unions, ``Literal``\ s, and other types. This kind
```

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,977 @@
+   trycast_type(MyTypedDict, dict(value='hi'))  # questionable; accepted by mypy 1.9.0
+
+To solve that problem, ``type[]`` could be widened to include the
+additional values allowed by ``TypeExpr``. However doing so would lose
```
```suggestion
additional values allowed by ``TypeExpr``. However, doing so would lose
```

davidfstr replied:
The pause introduced by the added comma breaks the flow when I read the paragraph. Leaving out.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,977 @@
+   -  Examples: ``int``, ``str``, ``MyClass``
+
+-  :ref:`Type expressions <typing:type-expression>`
+   include any type annotation which describes a type.
```
```suggestion
   include any expression which describes a type.
```

davidfstr replied:
Both the "Type expressions" and "Annotation expressions" sections start with "X include any type annotation", so I'd like to keep them consistent.

I think it would be misleading to make the paired change to the "Annotation expressions" section, which would then read "Annotation expressions include any **expression**, including [...]" and be too general a statement.

JelleZijlstra replied:
However, type expressions are not necessarily annotations; `T` in `cast(T, expr)` is a type expression.

JelleZijlstra commented on peps/pep-0747.rst:
```diff
@@ -0,0 +1,977 @@
+
+::
+
+   OPTIONAL_INT_TYPE: TypeExpr = TypeExpr[Optional[int]]  # OK
```
```suggestion
   OPTIONAL_INT_TYPE: TypeExpr = TypeExpr[int | None]  # OK
```

JelleZijlstra reviewed (commented):
Thanks, just a few small comments.

davidfstr committed:
Apply minor feedback

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

davidfstr commented:
(All feedback from Jelle has now either been applied to responded to)

davidfstr commented:
Thanks @JelleZijlstra , @hugovk , and @AlexWaygood for providing feedback!

I'll create a new discuss.python.org thread for this PEP within the next day or two.
